core/blockchain.go

```diff
@@ -1255,6 +1261,16 @@ func (bc *BlockChain) WriteBlockWithState(
 		return NonStatTy, err
 	}
 
+	if bc.chainConfig.IsEnablePruneBeaconChainFeature() {
+		nowBlockNum := block.Number().Uint64()
+		maxBlockNum := big.NewInt(0).Sub(block.Number(), big.NewInt(pruneBeaconChainBlockBefore)).Uint64()
+		maxEpoch := big.NewInt(0).Sub(block.Epoch(), big.NewInt(pruneBeaconChainBeforeEpoch))
+		if rand.Intn(100) == 1 && nowBlockNum > pruneBeaconChainBlockBefore && maxEpoch.Uint64() > pruneBeaconChainBeforeEpoch {
```

Comment:
why not prune 1 block for each new block inserted?

Reply:
These two conditions seems not quite reasonable "nowBlockNum > pruneBeaconChainBlockBefore && maxEpoch.Uint64() > pruneBeaconChainBeforeEpoch"

Reply:
And These two conditions seems not quite reasonable "nowBlockNum > pruneBeaconChainBlockBefore && maxEpoch.Uint64() > pruneBeaconChainBeforeEpoch"

Reply:
use `rand.Intn(100) == 1` has two advantages:
1. Each time the data of up to 10,000 keys is deleted, this is enough to keep the deletion to the latest with this probability. At the same time, it can reduce the number of IOs and minimize the impact, The Compact that contains the following uses a lower probability, which is the same reason.
2. If delete every new block added, this is equivalent to directly increasing the IO of the add new block, which will affect every node. If we use probabilization, it will at least not be at the same time.

This is prune 1 block for each new block, and compact every 10 blocks:
![image](https://user-images.githubusercontent.com/14296262/139830348-94449379-33ec-4a3e-9783-f673474cc496.png)

it used avg 23.9M/s disk write.

This is prune 1 block for every 100 blocks, and compact every 1000 blocks:
![image](https://user-images.githubusercontent.com/14296262/139830380-d31df867-1b87-4541-8dd5-20ff0c190c94.png)

it used avg 8.7M/s disk write.

Reply:
And These two conditions "nowBlockNum > pruneBeaconChainBlockBefore && maxEpoch.Uint64() > pruneBeaconChainBeforeEpoch", because `Uint64` Is unsigned, If a user does not use the snapshot to start a node,  Deletion will not start until 16 epoch are synchronized

Reply:
I don't get this "This is prune 1 block for every 100 blocks, and compact every 1000 blocks(in testnet shard 1):" shouldn't you prune 100 blocks for every 100 blocks? 

The difference here is whether you do small work for each block, or you batch all the work together every X blocks. IMO, spreading the work evenly for every block is less impactful to the node. But of course, you can prove me wrong with testing.

Reply:
The condition is confusing. Better just make it directly readable. Don't rely on uint overflow for condition, it's easily causing mistakes. For example, "maxEpoch.Uint64() > pruneBeaconChainBeforeEpoch" this one seems wrong. When maxEpoch is negative, it works. When it's positive. It means it is only true when maxEpoch > 16? Then block.Epoch() have to be 32?

Reply:
I corrected the conditions. and And I re-observed, most of the IO in compact stage. It's prune 1 block for each new block inserted now.